packages/gatsby/src/internal-plugins/functions/gatsby-node.ts

```diff
@@ -373,26 +428,29 @@ export async function onPreBootstrap({
             ],
             { ignoreInitial: true }
           )
-          .on(`all`, (event, path) => {
-            // Ignore change events from the API directory
-            if (event === `change` && path.includes(`/src/api/`)) {
+          .on(`all`, async (event, path) => {
+            // Ignore change events from the API directory for functions we're
+            // already watching.
+            if (
+              event === `change` &&
+              Object.values(activeEntries).includes(path) &&
+              path.includes(`/src/api/`)
+            ) {
               return
             }
 
             reporter.log(
               `Restarting function watcher due to change to "${path}"`
             )
 
-            // Otherwise, restart the watcher
-            compiler.close(async () => {
-              const config = await createWebpackConfig({
-                siteDirectoryPath,
-                functionsDirectory,
-                store,
-                reporter,
-              })
-              compiler = webpack(config).watch({}, callback)
+            const config = await createWebpackConfig({
+              siteDirectoryPath,
+              functionsDirectory,
+              store,
+              reporter,
             })
+
+            compiler = await restartWebpack(compiler, config, callback)
```

Comment:
Why do we need to restart the watcher?

Reply:
For two reasons:

1. an environment variable changes so we need to reset the `DefinePlugin` config
2. a user creates or deletes a function

Reply:
ah ok, good to know — it's my working assumption webpack can do anything if you spend enough hours looking for it but restarting was very fast (at least < 250ms) so hadn't worried about it.